---
title: "Test of ncbi_search backward compatibility"
author: "Zachary Foster"
date: "Tuesday, September 30, 2014"
output: html_document
---

### Updated code:

```{r}
#' Search for gene sequences available for a species from NCBI.
#'
#' @import XML httr stringr data.table
#' @template ncbi
#' @param id (character) Taxonomic id to search for. Not compatible with argument \code{taxa}.
#' @param limit Number of sequences to search for and return. Max of 10,000.
#'    If you search for 6000 records, and only 5000 are found, you will of course
#'    only get 5000 back.
#' @param entrez_query (\code{character}; length 1) An Entrez-format query to filter results with. 
#' @param hypothetical (\code{logical}; length 1) If \code{FALSE}, an attempt will be made to not
#'   return hypothetical or predicted sequences judging from accession number prefixs. 
#' @details Removes predicted sequences so you don't have to remove them.
#'     Predicted sequences are those with accession numbers that have "XM_" or
#'   	"XR_" prefixes.
#' @return Data.frame of results.
#' @seealso \code{\link[taxize]{ncbi_getbyid}}, \code{\link[taxize]{ncbi_getbyname}}
#' @author Scott Chamberlain \email{myrmecocystus@@gmail.com}, Zachary Foster 
#'   \email{zacharyfoster1989@@gmail.com}
#' @examples \donttest{
#' # A single species
#' out <- ncbi_search(taxa="Umbra limi", seqrange = "1:2000")
#' # Get the same species information using a taxonomy id 
#' out <- ncbi_search(id = "75935", seqrange = "1:2000")
#' # If the taxon name is unique, using the taxon name and id are equivalent
#' all(ncbi_search(id = "75935") ==  ncbi_search(taxa="Umbra limi"))
#' # If the taxon name is not unique, use taxon id
#' #  "266948" is the uid for the butterfly genus, but there is also a genus of orchids with the 
#' #  same name
#' nrow(ncbi_search(id = "266948")) ==  nrow(ncbi_search(taxa="Satyrium"))
#' # get list of genes available, removing non-unique
#' unique(out$gene_desc)
#' # does the string 'RAG1' exist in any of the gene names
#' out[grep("RAG1", out$gene_desc, ignore.case=TRUE),]
#'
#' # A single species without records in NCBI
#' out <- ncbi_search(taxa="Sequoia wellingtonia", seqrange="1:2000", getrelated=TRUE)
#'
#' # Many species, can run in parallel or not using plyr
#' species <- c("Salvelinus alpinus","Ictalurus nebulosus","Carassius auratus")
#' out2 <- ncbi_search(taxa=species, seqrange = "1:2000")
#' lapply(out2, head) # see heads of all
#' library("plyr")
#' out2df <- ldply(out2) # make data.frame of all
#' unique(out2df$gene_desc) # get list of genes available, removing non-unique
#' out2df[grep("60S ribosomal protein", out2df$gene_desc, ignore.case=TRUE),] # search across all
#' }
#' @export
ncbi_search2 <- function(taxa = NULL, id = NULL, seqrange="1:3000", getrelated=FALSE, limit = 500,
                        entrez_query = NULL, hypothetical = FALSE, verbose=TRUE)
{
  
  # Function to search for sequences with esearch
  search_for_sequences <- function(id) {
    # Contruct search query
    query_term <- paste0("xXarbitraryXx[porgn:__txid", id, "] AND ", seqrange, " [SLEN]")
    if (!is.null(entrez_query)) query_term <- paste(query_term, entrez_query, sep = " AND ")
    query <- list(db = "nuccore", retmax = limit, term = query_term)
    # Submit query to NCBI
    query_init <- GET(url_esearch, query=query)
    stop_for_status(query_init)
    # Parse result
    out <- xpathApply(content(query_init, as="parsed"), "//eSearchResult")[[1]]
    return(out)
  }

  # Function to parse results from http query
  parseres <- function(x){
    outsum <- xpathApply(content(x, as="parsed"), "//eSummaryResult")[[1]]
    names <- sapply(getNodeSet(outsum[[1]], "//Item"), xmlGetAttr, name="Name") # gets names of values in summary
    predicted <- as.character(sapply(getNodeSet(outsum, "//Item"), xmlValue)[str_detect(names, "Caption")]) #  get access numbers
    has_access_prefix <- grepl("_", predicted)
    access_prefix <- unlist(Map(function(x, y) ifelse(x, strsplit(y, "_")[[1]][[1]], NA),
                         has_access_prefix, predicted))
    predicted[has_access_prefix] <- vapply(strsplit(predicted[has_access_prefix], "_"), `[[`, character(1), 2)
    length_ <- as.numeric(sapply(getNodeSet(outsum, "//Item"), xmlValue)[str_detect(names, "Length")]) # gets seq lengths
    gis <- as.numeric(sapply(getNodeSet(outsum, "//Item"), xmlValue)[str_detect(names, "Gi")]) # gets GI numbers
    spnames <- sapply(getNodeSet(outsum, "//Item"), xmlValue)[str_detect(names, "Title")] # gets seq lengths # get spp names
    spused <- sapply(spnames, function(x) paste(str_split(x, " ")[[1]][1:2], sep="", collapse=" "), USE.NAMES=FALSE)
    genesavail <- sapply(spnames, function(x) paste(str_split(x, " ")[[1]][-c(1:2)], sep="", collapse=" "), USE.NAMES=FALSE)
    df <- data.frame(spused=spused, length=length_, genesavail=genesavail, access_num=predicted, ids=gis, stringsAsFactors=FALSE)
    if (!hypothetical) df <- df[!(access_prefix %in% c("XM","XR")),]
    return(df)
  }
  
  # Function to download sequence summary information using esummary
  download_summary <- function(seq_id) {
    actualnum <- length(seq_id)
    if(actualnum > 10000){
      q <- list(db = "nucleotide")
      getstart <- seq(from=1, to=actualnum, by=10000)
      getnum <- c(rep(10000, length(getstart)-1), actualnum-sum(rep(10000, length(getstart)-1)))
      iterlist = list()
      for(i in seq_along(getstart)){
        q$id = paste(seq_id[getstart[i]:(getstart[i]+(getnum[i]-1))], collapse=" ")
        q$retstart <- getstart[i]
        q$retmax <- getnum[i]
        query_res <- POST(url_esummary, body=q)
        stop_for_status(query_res)
        iterlist[[i]] <- parseres(query_res)
      }
      df <- data.frame(rbindlist(iterlist))
    } else
    {
      q <- list(db = "nucleotide", id = paste(seq_id, collapse=" "))
      query_res <- POST(url_esummary, body=q)
      stop_for_status(query_res)
      df <- parseres(query_res)
    }
    return(df)
  }
  
  # Constants
  url_esearch <- "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi"
  url_esummary <- "http://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi"

  # Argument validation 
  if (sum(c(is.null(taxa), is.null(id))) != 1) {
    stop("Either taxa or id must be specified, but not both")
  }
  if ((!is.null(id)) && getrelated) warning("option 'getrelated' has no effect when used with option 'id'.")

  # Convert 'taxa' to 'id' if 'taxa' is supplied
  if (is.null(id)) {
    id <- get_uid(taxa)
  } else {
    id <- as.character(id)
    class(id) <- "uid"
  }
  
  get_parent <- function(child) {
      query_classification <- classification(id = xx, db = "ncbi")[[1]]
      if (nrow(query_classification) > 1) {
        parent_name <- query_classification$name[nrow(query_classification) - 1]
        mssg(verbose, paste0("...retrieving sequence IDs for ", parent_name, "..."))
        xx <- get_uid(parent_name)
        
      } else {  
        mssg(verbose, paste0("no related taxa found"))        
      }    
  }
  
  # Function to process queries one at a time
  foo <- function(xx){
    if (is.na(xx)) return(NA)
    mssg(verbose, paste("Working on ", xx, "...", sep=""))
    mssg(verbose, "...retrieving sequence IDs...")
    esearch_result <- search_for_sequences(xx)
    if (as.numeric(xmlValue(xpathApply(esearch_result, "//Count")[[1]])) == 0 && getrelated) {
      mssg(verbose, paste("no sequences for ", xx, " - getting other sp.", sep=""))
      parent_id <- get_parent(xx)
      esearch_result <- search_for_sequences(parent_id)
    }
    if (as.numeric(xmlValue(xpathApply(esearch_result, "//Count")[[1]])) == 0) {
      mssg(verbose, "no sequences found")
      df <- data.frame(taxon = character(0), length = numeric(0), gene_desc = character(0),
                       acc_no = character(0), gi_no = numeric(0))
    } else {
      seq_ids <- xpathSApply(esearch_result, "//IdList//Id", xmlValue) # Get sequence IDs in list
      mssg(verbose, "...retrieving available genes and their lengths...")
      df <- download_summary(seq_ids)    
      mssg(verbose, "...done.")
   } 
    names(df) <- c("taxon","length","gene_desc","acc_no","gi_no")
    return(df)
  }
  
  foo_safe <- plyr::failwith(NULL, foo)
  if (length(id) == 1) {
    foo_safe(id)
  } else {
    lapply(id, foo_safe)
  }
}
```

```{r}
install.packages("taxize")
library(taxize)
library(testthat)
environment(ncbi_search2) <- asNamespace("taxize")
```


### Output format tests

#### Single query 

For a single query, both return a single data frame. Although this is convenient for people using the funtion directly, it complicates the use of the function by other functions, since the return format for multiple queries (list) is different from the return format for a single query. I preserved this behavior for backward compatibility, but I think returning a one-element list is more consistent. 

```{r}
query <- c("pythium")
x <- ncbi_search(taxa = query, limit = 50)
str(x)
y <- ncbi_search2(taxa = query, limit = 50)
str(y)
all.equal(x, y)
```

#### Multiple queries

```{r}
query <- c("pythium", "phytophthora")
x <- ncbi_search(taxa = query, limit = 50)
str(x)
y <- ncbi_search2(taxa = query, limit = 50)
str(y)
all.equal(x, y)
```

#### No results

The two functions differ in how they handle edge cases, like no results or invalid taxon names. My version of the function returns a 0-row dataframe if no sequences are found. This allows for the results of sucessful and unsucessful searches to be `rbind`ed together without further modifications. On the other hand, an invalid search term returns `NA`, since this is something the use probably did not intend.

```{r}
query <- c("fungi", "this_is_not_valid")
x <- ncbi_search(taxa = query, limit = 50)
str(x)
y <- ncbi_search2(taxa = query, limit = 50)
str(y)
isTRUE(all.equal(x, y))
```

An error during processing still returns `NULL` for that element, like it does in the current version of `ncbi_search`. This is a logical output, but there might be cases when a user is running a list of queries and `rbind`ing the result. In this case, the error could be overlooked, since there would be no record of it in the data. Im not sure what a better choice would be, so I left it as `NULL`.

```{r}
all.equal(rbind(NULL, data.frame(1:10)), data.frame(1:10))
```

### New functionality

#### `entrez_query` option

This option allows advanced users to add their own filter to the esearch query using entrez query syntax. 

```{r}
query <- c("pythium")
x <- ncbi_search2(taxa = query, limit = 50, entrez_query = "18S[title] AND 28S[title]")
head(x$gene_desc)
```

#### `hypothetical` option

This allows users to include hypothetical/predicted genes if they want. The default to remove them, as in the built-in functionality of the current `ncbi_search`. 

```{r}
query <- c("fungi")
x <- ncbi_search2(taxa = query, limit = 50)
str(x)
y <- ncbi_search2(taxa = query, limit = 50, hypothetical = TRUE)
str(y)
```

### Other changes

#### Handling of taxon names

If a taxon is supplied as a character vector via the `taxa` argument, it is converted to a ncbi taxonomy uid using `get_uid`. Instead of searching ncbi for the name, the taxa id is used. This simplifies the downstream analysis and uses `get_uid` to resolve ambiguities. 

```{r}
query <- c("Satyrium") #An ambiguous genus name
x <- ncbi_search(taxa = query, limit = 50)
str(x)
y <- ncbi_search2(taxa = query, limit = 50)
str(y)
isTRUE(all.equal(x, y))
```


#### Implementation of `getrelated` option

In the curent `ncbi_search` the `getrelated` option seems to work by extracting the genus name of a latin binomial name (it assumes the input is a species). I modified it to find its parent using `classification(x, db = " ncbi")`. This means that the `getrelated` option should now work at any taxonomic level. Since the `taxa` argument is converted to a uid, this change also means that the `getrelated` option will work wether an uid or a taxon name is given. 

```{r}
query <- c("Enhydrus grayii")
x <- ncbi_search(taxa = query, limit = 50, getrelated =  TRUE)
str(x)
y <- ncbi_search2(taxa = query, limit = 50, getrelated =  TRUE)
str(y)

```
