<!DOCTYPE html>
<!-- saved from url=(0014)about:internet -->
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Updated code:</title>

<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 12px;
   margin: 8px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 { 
   font-size:2.2em; 
}

h2 { 
   font-size:1.8em; 
}

h3 { 
   font-size:1.4em; 
}

h4 { 
   font-size:1.0em; 
}

h5 { 
   font-size:0.9em; 
}

h6 { 
   font-size:0.8em; 
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre {	
   margin-top: 0;
   max-width: 95%;
   border: 1px solid #ccc;
   white-space: pre-wrap;
}

pre code {
   display: block; padding: 0.5em;
}

code.r, code.cpp {
   background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * { 
      background: transparent !important; 
      color: black !important; 
      filter:none !important; 
      -ms-filter: none !important; 
   }

   body { 
      font-size:12pt; 
      max-width:100%; 
   }
       
   a, a:visited { 
      text-decoration: underline; 
   }

   hr { 
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote { 
      padding-right: 1em; 
      page-break-inside: avoid; 
   }

   tr, img { 
      page-break-inside: avoid; 
   }

   img { 
      max-width: 100% !important; 
   }

   @page :left { 
      margin: 15mm 20mm 15mm 10mm; 
   }
     
   @page :right { 
      margin: 15mm 10mm 15mm 20mm; 
   }

   p, h2, h3 { 
      orphans: 3; widows: 3; 
   }

   h2, h3 { 
      page-break-after: avoid; 
   }
}

</style>

<!-- Styles for R syntax highlighter -->
<style type="text/css">
   pre .operator,
   pre .paren {
     color: rgb(104, 118, 135)
   }

   pre .literal {
     color: rgb(88, 72, 246)
   }

   pre .number {
     color: rgb(0, 0, 205);
   }

   pre .comment {
     color: rgb(76, 136, 107);
   }

   pre .keyword {
     color: rgb(0, 0, 255);
   }

   pre .identifier {
     color: rgb(0, 0, 0);
   }

   pre .string {
     color: rgb(3, 106, 7);
   }
</style>

<!-- R syntax highlighter -->
<script type="text/javascript">
var hljs=new function(){function m(p){return p.replace(/&/gm,"&amp;").replace(/</gm,"&lt;")}function f(r,q,p){return RegExp(q,"m"+(r.cI?"i":"")+(p?"g":""))}function b(r){for(var p=0;p<r.childNodes.length;p++){var q=r.childNodes[p];if(q.nodeName=="CODE"){return q}if(!(q.nodeType==3&&q.nodeValue.match(/\s+/))){break}}}function h(t,s){var p="";for(var r=0;r<t.childNodes.length;r++){if(t.childNodes[r].nodeType==3){var q=t.childNodes[r].nodeValue;if(s){q=q.replace(/\n/g,"")}p+=q}else{if(t.childNodes[r].nodeName=="BR"){p+="\n"}else{p+=h(t.childNodes[r])}}}if(/MSIE [678]/.test(navigator.userAgent)){p=p.replace(/\r/g,"\n")}return p}function a(s){var r=s.className.split(/\s+/);r=r.concat(s.parentNode.className.split(/\s+/));for(var q=0;q<r.length;q++){var p=r[q].replace(/^language-/,"");if(e[p]){return p}}}function c(q){var p=[];(function(s,t){for(var r=0;r<s.childNodes.length;r++){if(s.childNodes[r].nodeType==3){t+=s.childNodes[r].nodeValue.length}else{if(s.childNodes[r].nodeName=="BR"){t+=1}else{if(s.childNodes[r].nodeType==1){p.push({event:"start",offset:t,node:s.childNodes[r]});t=arguments.callee(s.childNodes[r],t);p.push({event:"stop",offset:t,node:s.childNodes[r]})}}}}return t})(q,0);return p}function k(y,w,x){var q=0;var z="";var s=[];function u(){if(y.length&&w.length){if(y[0].offset!=w[0].offset){return(y[0].offset<w[0].offset)?y:w}else{return w[0].event=="start"?y:w}}else{return y.length?y:w}}function t(D){var A="<"+D.nodeName.toLowerCase();for(var B=0;B<D.attributes.length;B++){var C=D.attributes[B];A+=" "+C.nodeName.toLowerCase();if(C.value!==undefined&&C.value!==false&&C.value!==null){A+='="'+m(C.value)+'"'}}return A+">"}while(y.length||w.length){var v=u().splice(0,1)[0];z+=m(x.substr(q,v.offset-q));q=v.offset;if(v.event=="start"){z+=t(v.node);s.push(v.node)}else{if(v.event=="stop"){var p,r=s.length;do{r--;p=s[r];z+=("</"+p.nodeName.toLowerCase()+">")}while(p!=v.node);s.splice(r,1);while(r<s.length){z+=t(s[r]);r++}}}}return z+m(x.substr(q))}function j(){function q(x,y,v){if(x.compiled){return}var u;var s=[];if(x.k){x.lR=f(y,x.l||hljs.IR,true);for(var w in x.k){if(!x.k.hasOwnProperty(w)){continue}if(x.k[w] instanceof Object){u=x.k[w]}else{u=x.k;w="keyword"}for(var r in u){if(!u.hasOwnProperty(r)){continue}x.k[r]=[w,u[r]];s.push(r)}}}if(!v){if(x.bWK){x.b="\\b("+s.join("|")+")\\s"}x.bR=f(y,x.b?x.b:"\\B|\\b");if(!x.e&&!x.eW){x.e="\\B|\\b"}if(x.e){x.eR=f(y,x.e)}}if(x.i){x.iR=f(y,x.i)}if(x.r===undefined){x.r=1}if(!x.c){x.c=[]}x.compiled=true;for(var t=0;t<x.c.length;t++){if(x.c[t]=="self"){x.c[t]=x}q(x.c[t],y,false)}if(x.starts){q(x.starts,y,false)}}for(var p in e){if(!e.hasOwnProperty(p)){continue}q(e[p].dM,e[p],true)}}function d(B,C){if(!j.called){j();j.called=true}function q(r,M){for(var L=0;L<M.c.length;L++){if((M.c[L].bR.exec(r)||[null])[0]==r){return M.c[L]}}}function v(L,r){if(D[L].e&&D[L].eR.test(r)){return 1}if(D[L].eW){var M=v(L-1,r);return M?M+1:0}return 0}function w(r,L){return L.i&&L.iR.test(r)}function K(N,O){var M=[];for(var L=0;L<N.c.length;L++){M.push(N.c[L].b)}var r=D.length-1;do{if(D[r].e){M.push(D[r].e)}r--}while(D[r+1].eW);if(N.i){M.push(N.i)}return f(O,M.join("|"),true)}function p(M,L){var N=D[D.length-1];if(!N.t){N.t=K(N,E)}N.t.lastIndex=L;var r=N.t.exec(M);return r?[M.substr(L,r.index-L),r[0],false]:[M.substr(L),"",true]}function z(N,r){var L=E.cI?r[0].toLowerCase():r[0];var M=N.k[L];if(M&&M instanceof Array){return M}return false}function F(L,P){L=m(L);if(!P.k){return L}var r="";var O=0;P.lR.lastIndex=0;var M=P.lR.exec(L);while(M){r+=L.substr(O,M.index-O);var N=z(P,M);if(N){x+=N[1];r+='<span class="'+N[0]+'">'+M[0]+"</span>"}else{r+=M[0]}O=P.lR.lastIndex;M=P.lR.exec(L)}return r+L.substr(O,L.length-O)}function J(L,M){if(M.sL&&e[M.sL]){var r=d(M.sL,L);x+=r.keyword_count;return r.value}else{return F(L,M)}}function I(M,r){var L=M.cN?'<span class="'+M.cN+'">':"";if(M.rB){y+=L;M.buffer=""}else{if(M.eB){y+=m(r)+L;M.buffer=""}else{y+=L;M.buffer=r}}D.push(M);A+=M.r}function G(N,M,Q){var R=D[D.length-1];if(Q){y+=J(R.buffer+N,R);return false}var P=q(M,R);if(P){y+=J(R.buffer+N,R);I(P,M);return P.rB}var L=v(D.length-1,M);if(L){var O=R.cN?"</span>":"";if(R.rE){y+=J(R.buffer+N,R)+O}else{if(R.eE){y+=J(R.buffer+N,R)+O+m(M)}else{y+=J(R.buffer+N+M,R)+O}}while(L>1){O=D[D.length-2].cN?"</span>":"";y+=O;L--;D.length--}var r=D[D.length-1];D.length--;D[D.length-1].buffer="";if(r.starts){I(r.starts,"")}return R.rE}if(w(M,R)){throw"Illegal"}}var E=e[B];var D=[E.dM];var A=0;var x=0;var y="";try{var s,u=0;E.dM.buffer="";do{s=p(C,u);var t=G(s[0],s[1],s[2]);u+=s[0].length;if(!t){u+=s[1].length}}while(!s[2]);if(D.length>1){throw"Illegal"}return{r:A,keyword_count:x,value:y}}catch(H){if(H=="Illegal"){return{r:0,keyword_count:0,value:m(C)}}else{throw H}}}function g(t){var p={keyword_count:0,r:0,value:m(t)};var r=p;for(var q in e){if(!e.hasOwnProperty(q)){continue}var s=d(q,t);s.language=q;if(s.keyword_count+s.r>r.keyword_count+r.r){r=s}if(s.keyword_count+s.r>p.keyword_count+p.r){r=p;p=s}}if(r.language){p.second_best=r}return p}function i(r,q,p){if(q){r=r.replace(/^((<[^>]+>|\t)+)/gm,function(t,w,v,u){return w.replace(/\t/g,q)})}if(p){r=r.replace(/\n/g,"<br>")}return r}function n(t,w,r){var x=h(t,r);var v=a(t);var y,s;if(v){y=d(v,x)}else{return}var q=c(t);if(q.length){s=document.createElement("pre");s.innerHTML=y.value;y.value=k(q,c(s),x)}y.value=i(y.value,w,r);var u=t.className;if(!u.match("(\\s|^)(language-)?"+v+"(\\s|$)")){u=u?(u+" "+v):v}if(/MSIE [678]/.test(navigator.userAgent)&&t.tagName=="CODE"&&t.parentNode.tagName=="PRE"){s=t.parentNode;var p=document.createElement("div");p.innerHTML="<pre><code>"+y.value+"</code></pre>";t=p.firstChild.firstChild;p.firstChild.cN=s.cN;s.parentNode.replaceChild(p.firstChild,s)}else{t.innerHTML=y.value}t.className=u;t.result={language:v,kw:y.keyword_count,re:y.r};if(y.second_best){t.second_best={language:y.second_best.language,kw:y.second_best.keyword_count,re:y.second_best.r}}}function o(){if(o.called){return}o.called=true;var r=document.getElementsByTagName("pre");for(var p=0;p<r.length;p++){var q=b(r[p]);if(q){n(q,hljs.tabReplace)}}}function l(){if(window.addEventListener){window.addEventListener("DOMContentLoaded",o,false);window.addEventListener("load",o,false)}else{if(window.attachEvent){window.attachEvent("onload",o)}else{window.onload=o}}}var e={};this.LANGUAGES=e;this.highlight=d;this.highlightAuto=g;this.fixMarkup=i;this.highlightBlock=n;this.initHighlighting=o;this.initHighlightingOnLoad=l;this.IR="[a-zA-Z][a-zA-Z0-9_]*";this.UIR="[a-zA-Z_][a-zA-Z0-9_]*";this.NR="\\b\\d+(\\.\\d+)?";this.CNR="\\b(0[xX][a-fA-F0-9]+|(\\d+(\\.\\d*)?|\\.\\d+)([eE][-+]?\\d+)?)";this.BNR="\\b(0b[01]+)";this.RSR="!|!=|!==|%|%=|&|&&|&=|\\*|\\*=|\\+|\\+=|,|\\.|-|-=|/|/=|:|;|<|<<|<<=|<=|=|==|===|>|>=|>>|>>=|>>>|>>>=|\\?|\\[|\\{|\\(|\\^|\\^=|\\||\\|=|\\|\\||~";this.ER="(?![\\s\\S])";this.BE={b:"\\\\.",r:0};this.ASM={cN:"string",b:"'",e:"'",i:"\\n",c:[this.BE],r:0};this.QSM={cN:"string",b:'"',e:'"',i:"\\n",c:[this.BE],r:0};this.CLCM={cN:"comment",b:"//",e:"$"};this.CBLCLM={cN:"comment",b:"/\\*",e:"\\*/"};this.HCM={cN:"comment",b:"#",e:"$"};this.NM={cN:"number",b:this.NR,r:0};this.CNM={cN:"number",b:this.CNR,r:0};this.BNM={cN:"number",b:this.BNR,r:0};this.inherit=function(r,s){var p={};for(var q in r){p[q]=r[q]}if(s){for(var q in s){p[q]=s[q]}}return p}}();hljs.LANGUAGES.cpp=function(){var a={keyword:{"false":1,"int":1,"float":1,"while":1,"private":1,"char":1,"catch":1,"export":1,virtual:1,operator:2,sizeof:2,dynamic_cast:2,typedef:2,const_cast:2,"const":1,struct:1,"for":1,static_cast:2,union:1,namespace:1,unsigned:1,"long":1,"throw":1,"volatile":2,"static":1,"protected":1,bool:1,template:1,mutable:1,"if":1,"public":1,friend:2,"do":1,"return":1,"goto":1,auto:1,"void":2,"enum":1,"else":1,"break":1,"new":1,extern:1,using:1,"true":1,"class":1,asm:1,"case":1,typeid:1,"short":1,reinterpret_cast:2,"default":1,"double":1,register:1,explicit:1,signed:1,typename:1,"try":1,"this":1,"switch":1,"continue":1,wchar_t:1,inline:1,"delete":1,alignof:1,char16_t:1,char32_t:1,constexpr:1,decltype:1,noexcept:1,nullptr:1,static_assert:1,thread_local:1,restrict:1,_Bool:1,complex:1},built_in:{std:1,string:1,cin:1,cout:1,cerr:1,clog:1,stringstream:1,istringstream:1,ostringstream:1,auto_ptr:1,deque:1,list:1,queue:1,stack:1,vector:1,map:1,set:1,bitset:1,multiset:1,multimap:1,unordered_set:1,unordered_map:1,unordered_multiset:1,unordered_multimap:1,array:1,shared_ptr:1}};return{dM:{k:a,i:"</",c:[hljs.CLCM,hljs.CBLCLM,hljs.QSM,{cN:"string",b:"'\\\\?.",e:"'",i:"."},{cN:"number",b:"\\b(\\d+(\\.\\d*)?|\\.\\d+)(u|U|l|L|ul|UL|f|F)"},hljs.CNM,{cN:"preprocessor",b:"#",e:"$"},{cN:"stl_container",b:"\\b(deque|list|queue|stack|vector|map|set|bitset|multiset|multimap|unordered_map|unordered_set|unordered_multiset|unordered_multimap|array)\\s*<",e:">",k:a,r:10,c:["self"]}]}}}();hljs.LANGUAGES.r={dM:{c:[hljs.HCM,{cN:"number",b:"\\b0[xX][0-9a-fA-F]+[Li]?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+(?:[eE][+\\-]?\\d*)?L\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\b\\d+\\.(?!\\d)(?:i\\b)?",e:hljs.IMMEDIATE_RE,r:1},{cN:"number",b:"\\b\\d+(?:\\.\\d*)?(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"number",b:"\\.\\d+(?:[eE][+\\-]?\\d*)?i?\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"keyword",b:"(?:tryCatch|library|setGeneric|setGroupGeneric)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\.",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\.\\.\\d+(?![\\w.])",e:hljs.IMMEDIATE_RE,r:10},{cN:"keyword",b:"\\b(?:function)",e:hljs.IMMEDIATE_RE,r:2},{cN:"keyword",b:"(?:if|in|break|next|repeat|else|for|return|switch|while|try|stop|warning|require|attach|detach|source|setMethod|setClass)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"literal",b:"(?:NA|NA_integer_|NA_real_|NA_character_|NA_complex_)\\b",e:hljs.IMMEDIATE_RE,r:10},{cN:"literal",b:"(?:NULL|TRUE|FALSE|T|F|Inf|NaN)\\b",e:hljs.IMMEDIATE_RE,r:1},{cN:"identifier",b:"[a-zA-Z.][a-zA-Z0-9._]*\\b",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"<\\-(?!\\s*\\d)",e:hljs.IMMEDIATE_RE,r:2},{cN:"operator",b:"\\->|<\\-",e:hljs.IMMEDIATE_RE,r:1},{cN:"operator",b:"%%|~",e:hljs.IMMEDIATE_RE},{cN:"operator",b:">=|<=|==|!=|\\|\\||&&|=|\\+|\\-|\\*|/|\\^|>|<|!|&|\\||\\$|:",e:hljs.IMMEDIATE_RE,r:0},{cN:"operator",b:"%",e:"%",i:"\\n",r:1},{cN:"identifier",b:"`",e:"`",r:0},{cN:"string",b:'"',e:'"',c:[hljs.BE],r:0},{cN:"string",b:"'",e:"'",c:[hljs.BE],r:0},{cN:"paren",b:"[[({\\])}]",e:hljs.IMMEDIATE_RE,r:0}]}};
hljs.initHighlightingOnLoad();
</script>




</head>

<body>
<h3>Updated code:</h3>

<pre><code class="r">#&#39; Search for gene sequences available for a species from NCBI.
#&#39;
#&#39; @import XML httr stringr data.table
#&#39; @template ncbi
#&#39; @param id (character) Taxonomic id to search for. Not compatible with argument \code{taxa}.
#&#39; @param limit Number of sequences to search for and return. Max of 10,000.
#&#39;    If you search for 6000 records, and only 5000 are found, you will of course
#&#39;    only get 5000 back.
#&#39; @param entrez_query (\code{character}; length 1) An Entrez-format query to filter results with. 
#&#39; @param hypothetical (\code{logical}; length 1) If \code{FALSE}, an attempt will be made to not
#&#39;   return hypothetical or predicted sequences judging from accession number prefixs. 
#&#39; @details Removes predicted sequences so you don&#39;t have to remove them.
#&#39;     Predicted sequences are those with accession numbers that have &quot;XM_&quot; or
#&#39;      &quot;XR_&quot; prefixes.
#&#39; @return Data.frame of results.
#&#39; @seealso \code{\link[taxize]{ncbi_getbyid}}, \code{\link[taxize]{ncbi_getbyname}}
#&#39; @author Scott Chamberlain \email{myrmecocystus@@gmail.com}, Zachary Foster 
#&#39;   \email{zacharyfoster1989@@gmail.com}
#&#39; @examples \donttest{
#&#39; # A single species
#&#39; out &lt;- ncbi_search(taxa=&quot;Umbra limi&quot;, seqrange = &quot;1:2000&quot;)
#&#39; # Get the same species information using a taxonomy id 
#&#39; out &lt;- ncbi_search(id = &quot;75935&quot;, seqrange = &quot;1:2000&quot;)
#&#39; # If the taxon name is unique, using the taxon name and id are equivalent
#&#39; all(ncbi_search(id = &quot;75935&quot;) ==  ncbi_search(taxa=&quot;Umbra limi&quot;))
#&#39; # If the taxon name is not unique, use taxon id
#&#39; #  &quot;266948&quot; is the uid for the butterfly genus, but there is also a genus of orchids with the 
#&#39; #  same name
#&#39; nrow(ncbi_search(id = &quot;266948&quot;)) ==  nrow(ncbi_search(taxa=&quot;Satyrium&quot;))
#&#39; # get list of genes available, removing non-unique
#&#39; unique(out$gene_desc)
#&#39; # does the string &#39;RAG1&#39; exist in any of the gene names
#&#39; out[grep(&quot;RAG1&quot;, out$gene_desc, ignore.case=TRUE),]
#&#39;
#&#39; # A single species without records in NCBI
#&#39; out &lt;- ncbi_search(taxa=&quot;Sequoia wellingtonia&quot;, seqrange=&quot;1:2000&quot;, getrelated=TRUE)
#&#39;
#&#39; # Many species, can run in parallel or not using plyr
#&#39; species &lt;- c(&quot;Salvelinus alpinus&quot;,&quot;Ictalurus nebulosus&quot;,&quot;Carassius auratus&quot;)
#&#39; out2 &lt;- ncbi_search(taxa=species, seqrange = &quot;1:2000&quot;)
#&#39; lapply(out2, head) # see heads of all
#&#39; library(&quot;plyr&quot;)
#&#39; out2df &lt;- ldply(out2) # make data.frame of all
#&#39; unique(out2df$gene_desc) # get list of genes available, removing non-unique
#&#39; out2df[grep(&quot;60S ribosomal protein&quot;, out2df$gene_desc, ignore.case=TRUE),] # search across all
#&#39; }
#&#39; @export
ncbi_search2 &lt;- function(taxa = NULL, id = NULL, seqrange=&quot;1:3000&quot;, getrelated=FALSE, limit = 500,
                        entrez_query = NULL, hypothetical = FALSE, verbose=TRUE)
{

  # Function to search for sequences with esearch
  search_for_sequences &lt;- function(id) {
    # Contruct search query
    query_term &lt;- paste0(&quot;xXarbitraryXx[porgn:__txid&quot;, id, &quot;] AND &quot;, seqrange, &quot; [SLEN]&quot;)
    if (!is.null(entrez_query)) query_term &lt;- paste(query_term, entrez_query, sep = &quot; AND &quot;)
    query &lt;- list(db = &quot;nuccore&quot;, retmax = limit, term = query_term)
    # Submit query to NCBI
    query_init &lt;- GET(url_esearch, query=query)
    stop_for_status(query_init)
    # Parse result
    out &lt;- xpathApply(content(query_init, as=&quot;parsed&quot;), &quot;//eSearchResult&quot;)[[1]]
    return(out)
  }

  # Function to parse results from http query
  parseres &lt;- function(x){
    outsum &lt;- xpathApply(content(x, as=&quot;parsed&quot;), &quot;//eSummaryResult&quot;)[[1]]
    names &lt;- sapply(getNodeSet(outsum[[1]], &quot;//Item&quot;), xmlGetAttr, name=&quot;Name&quot;) # gets names of values in summary
    predicted &lt;- as.character(sapply(getNodeSet(outsum, &quot;//Item&quot;), xmlValue)[str_detect(names, &quot;Caption&quot;)]) #  get access numbers
    has_access_prefix &lt;- grepl(&quot;_&quot;, predicted)
    access_prefix &lt;- unlist(Map(function(x, y) ifelse(x, strsplit(y, &quot;_&quot;)[[1]][[1]], NA),
                         has_access_prefix, predicted))
    predicted[has_access_prefix] &lt;- vapply(strsplit(predicted[has_access_prefix], &quot;_&quot;), `[[`, character(1), 2)
    length_ &lt;- as.numeric(sapply(getNodeSet(outsum, &quot;//Item&quot;), xmlValue)[str_detect(names, &quot;Length&quot;)]) # gets seq lengths
    gis &lt;- as.numeric(sapply(getNodeSet(outsum, &quot;//Item&quot;), xmlValue)[str_detect(names, &quot;Gi&quot;)]) # gets GI numbers
    spnames &lt;- sapply(getNodeSet(outsum, &quot;//Item&quot;), xmlValue)[str_detect(names, &quot;Title&quot;)] # gets seq lengths # get spp names
    spused &lt;- sapply(spnames, function(x) paste(str_split(x, &quot; &quot;)[[1]][1:2], sep=&quot;&quot;, collapse=&quot; &quot;), USE.NAMES=FALSE)
    genesavail &lt;- sapply(spnames, function(x) paste(str_split(x, &quot; &quot;)[[1]][-c(1:2)], sep=&quot;&quot;, collapse=&quot; &quot;), USE.NAMES=FALSE)
    df &lt;- data.frame(spused=spused, length=length_, genesavail=genesavail, access_num=predicted, ids=gis, stringsAsFactors=FALSE)
    if (!hypothetical) df &lt;- df[!(access_prefix %in% c(&quot;XM&quot;,&quot;XR&quot;)),]
    return(df)
  }

  # Function to download sequence summary information using esummary
  download_summary &lt;- function(seq_id) {
    actualnum &lt;- length(seq_id)
    if(actualnum &gt; 10000){
      q &lt;- list(db = &quot;nucleotide&quot;)
      getstart &lt;- seq(from=1, to=actualnum, by=10000)
      getnum &lt;- c(rep(10000, length(getstart)-1), actualnum-sum(rep(10000, length(getstart)-1)))
      iterlist = list()
      for(i in seq_along(getstart)){
        q$id = paste(seq_id[getstart[i]:(getstart[i]+(getnum[i]-1))], collapse=&quot; &quot;)
        q$retstart &lt;- getstart[i]
        q$retmax &lt;- getnum[i]
        query_res &lt;- POST(url_esummary, body=q)
        stop_for_status(query_res)
        iterlist[[i]] &lt;- parseres(query_res)
      }
      df &lt;- data.frame(rbindlist(iterlist))
    } else
    {
      q &lt;- list(db = &quot;nucleotide&quot;, id = paste(seq_id, collapse=&quot; &quot;))
      query_res &lt;- POST(url_esummary, body=q)
      stop_for_status(query_res)
      df &lt;- parseres(query_res)
    }
    return(df)
  }

  # Constants
  url_esearch &lt;- &quot;http://eutils.ncbi.nlm.nih.gov/entrez/eutils/esearch.fcgi&quot;
  url_esummary &lt;- &quot;http://eutils.ncbi.nlm.nih.gov/entrez/eutils/esummary.fcgi&quot;

  # Argument validation 
  if (sum(c(is.null(taxa), is.null(id))) != 1) {
    stop(&quot;Either taxa or id must be specified, but not both&quot;)
  }
  if ((!is.null(id)) &amp;&amp; getrelated) warning(&quot;option &#39;getrelated&#39; has no effect when used with option &#39;id&#39;.&quot;)

  # Convert &#39;taxa&#39; to &#39;id&#39; if &#39;taxa&#39; is supplied
  if (is.null(id)) {
    id &lt;- get_uid(taxa)
  } else {
    id &lt;- as.character(id)
    class(id) &lt;- &quot;uid&quot;
  }

  # Function to process queries one at a time
  foo &lt;- function(xx){
    if (is.na(xx)) return(NA)
    mssg(verbose, paste(&quot;Working on &quot;, xx, &quot;...&quot;, sep=&quot;&quot;))
    mssg(verbose, &quot;...retrieving sequence IDs...&quot;)
    esearch_result &lt;- search_for_sequences(xx)
    if (as.numeric(xmlValue(xpathApply(esearch_result, &quot;//Count&quot;)[[1]])) == 0 &amp;&amp; getrelated) {
      mssg(verbose, paste(&quot;no sequences for &quot;, xx, &quot; - getting other sp.&quot;, sep=&quot;&quot;))
      query_classification &lt;- classification(id = xx, db = &quot;ncbi&quot;)[[1]]
      if (nrow(query_classification) &gt; 1) {
        parent_name &lt;- query_classification$name[nrow(query_classification) - 1]
        mssg(verbose, paste0(&quot;...retrieving sequence IDs for &quot;, parent_name, &quot;...&quot;))
        xx &lt;- get_uid(parent_name)
        esearch_result &lt;- search_for_sequences(parent_id)
      } else {
        mssg(verbose, paste0(&quot;no related taxa found&quot;))        
      }
    }
    if (as.numeric(xmlValue(xpathApply(esearch_result, &quot;//Count&quot;)[[1]])) == 0) {
      mssg(verbose, &quot;no sequences found&quot;)
      df &lt;- data.frame(taxon = character(0), length = numeric(0), gene_desc = character(0),
                       acc_no = character(0), gi_no = numeric(0))
    } else {
      seq_ids &lt;- xpathSApply(esearch_result, &quot;//IdList//Id&quot;, xmlValue) # Get sequence IDs in list
      mssg(verbose, &quot;...retrieving available genes and their lengths...&quot;)
      df &lt;- download_summary(seq_ids)    
      mssg(verbose, &quot;...done.&quot;)
   } 
    names(df) &lt;- c(&quot;taxon&quot;,&quot;length&quot;,&quot;gene_desc&quot;,&quot;acc_no&quot;,&quot;gi_no&quot;)
    return(df)
  }

  foo_safe &lt;- plyr::failwith(NULL, foo)
  if (length(id) == 1) {
    foo_safe(id)
  } else {
    lapply(id, foo_safe)
  }
}
</code></pre>

<pre><code class="r">install.packages(&quot;taxize&quot;)
</code></pre>

<pre><code>## Installing package into &#39;/home/local/USDA-ARS/fosterz/R/x86_64-pc-linux-gnu-library/3.1&#39;
## (as &#39;lib&#39; is unspecified)
</code></pre>

<pre><code>## Error: trying to use CRAN without setting a mirror
</code></pre>

<pre><code class="r">library(taxize)
</code></pre>

<pre><code>## 
## 
## New to taxize? Tutorial at http://ropensci.org/tutorials/taxize_tutorial.html 
## citation(package=&#39;taxize&#39;) for the citation for this package 
## API key names have changed. Use tropicosApiKey, eolApiKey, ubioApiKey, and pmApiKey in your .Rprofile file. 
## Use suppressPackageStartupMessages() to suppress these startup messages in the future
</code></pre>

<pre><code class="r">library(testthat)
environment(ncbi_search2) &lt;- asNamespace(&quot;taxize&quot;)
</code></pre>

<h3>Output format tests</h3>

<h4>Single query</h4>

<p>For a single query, both return a single data frame. Although this is convenient for people using the funtion directly, it complicates the use of the function by other functions, since the return format for multiple queries (list) is different from the return format for a single query. I preserved this behavior for backward compatibility, but I think returning a one-element list is more consistent. </p>

<pre><code class="r">query &lt;- c(&quot;pythium&quot;)
x &lt;- ncbi_search(taxa = query, limit = 50)
</code></pre>

<pre><code>## Working on pythium...
## ...retrieving sequence IDs...
## ...retrieving available genes and their lengths...
## ...done.
</code></pre>

<pre><code class="r">str(x)
</code></pre>

<pre><code>## &#39;data.frame&#39;:    50 obs. of  5 variables:
##  $ taxon    : chr  &quot;Pythium dissotocum&quot; &quot;Pythium dissotocum&quot; &quot;Pythium dissotocum&quot; &quot;Pythium dissotocum&quot; ...
##  $ length   : num  514 861 871 828 862 862 856 874 894 785 ...
##  $ gene_desc: chr  &quot;voucher ABDN358 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1 and 5.8S ribosomal RNA gene, complete s&quot;| __truncated__ &quot;voucher ABDN356 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1, 5.8S ribosomal RNA gene, and internal &quot;| __truncated__ &quot;voucher ABDN355 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1, 5.8S ribosomal RNA gene, and internal &quot;| __truncated__ &quot;voucher ABDN354 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1 and 5.8S ribosomal RNA gene, complete s&quot;| __truncated__ ...
##  $ acc_no   : chr  &quot;KM061708&quot; &quot;KM061706&quot; &quot;KM061705&quot; &quot;KM061704&quot; ...
##  $ gi_no    : num  6.86e+08 6.86e+08 6.86e+08 6.86e+08 6.86e+08 ...
</code></pre>

<pre><code class="r">y &lt;- ncbi_search2(taxa = query, limit = 50)
</code></pre>

<pre><code>## 
## Retrieving data for taxon &#39;pythium&#39;
## 
## Working on 4797...
## ...retrieving sequence IDs...
## ...retrieving available genes and their lengths...
## ...done.
</code></pre>

<pre><code class="r">str(y)
</code></pre>

<pre><code>## &#39;data.frame&#39;:    50 obs. of  5 variables:
##  $ taxon    : chr  &quot;Pythium dissotocum&quot; &quot;Pythium dissotocum&quot; &quot;Pythium dissotocum&quot; &quot;Pythium dissotocum&quot; ...
##  $ length   : num  514 861 871 828 862 862 856 874 894 785 ...
##  $ gene_desc: chr  &quot;voucher ABDN358 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1 and 5.8S ribosomal RNA gene, complete s&quot;| __truncated__ &quot;voucher ABDN356 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1, 5.8S ribosomal RNA gene, and internal &quot;| __truncated__ &quot;voucher ABDN355 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1, 5.8S ribosomal RNA gene, and internal &quot;| __truncated__ &quot;voucher ABDN354 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1 and 5.8S ribosomal RNA gene, complete s&quot;| __truncated__ ...
##  $ acc_no   : chr  &quot;KM061708&quot; &quot;KM061706&quot; &quot;KM061705&quot; &quot;KM061704&quot; ...
##  $ gi_no    : num  6.86e+08 6.86e+08 6.86e+08 6.86e+08 6.86e+08 ...
</code></pre>

<pre><code class="r">all.equal(x, y)
</code></pre>

<pre><code>## [1] TRUE
</code></pre>

<h4>Multiple queries</h4>

<pre><code class="r">query &lt;- c(&quot;pythium&quot;, &quot;phytophthora&quot;)
x &lt;- ncbi_search(taxa = query, limit = 50)
</code></pre>

<pre><code>## Working on pythium...
## ...retrieving sequence IDs...
## ...retrieving available genes and their lengths...
## ...done.
## Working on phytophthora...
## ...retrieving sequence IDs...
## ...retrieving available genes and their lengths...
## ...done.
</code></pre>

<pre><code class="r">str(x)
</code></pre>

<pre><code>## List of 2
##  $ :&#39;data.frame&#39;:    50 obs. of  5 variables:
##   ..$ taxon    : chr [1:50] &quot;Pythium dissotocum&quot; &quot;Pythium dissotocum&quot; &quot;Pythium dissotocum&quot; &quot;Pythium dissotocum&quot; ...
##   ..$ length   : num [1:50] 514 861 871 828 862 862 856 874 894 785 ...
##   ..$ gene_desc: chr [1:50] &quot;voucher ABDN358 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1 and 5.8S ribosomal RNA gene, complete s&quot;| __truncated__ &quot;voucher ABDN356 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1, 5.8S ribosomal RNA gene, and internal &quot;| __truncated__ &quot;voucher ABDN355 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1, 5.8S ribosomal RNA gene, and internal &quot;| __truncated__ &quot;voucher ABDN354 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1 and 5.8S ribosomal RNA gene, complete s&quot;| __truncated__ ...
##   ..$ acc_no   : chr [1:50] &quot;KM061708&quot; &quot;KM061706&quot; &quot;KM061705&quot; &quot;KM061704&quot; ...
##   ..$ gi_no    : num [1:50] 6.86e+08 6.86e+08 6.86e+08 6.86e+08 6.86e+08 ...
##  $ :&#39;data.frame&#39;:    50 obs. of  5 variables:
##   ..$ taxon    : chr [1:50] &quot;Phytophthora humicola&quot; &quot;Phytophthora multivora&quot; &quot;Phytophthora multivora&quot; &quot;Phytophthora citrophthora&quot; ...
##   ..$ length   : num [1:50] 899 813 860 840 833 846 874 875 972 782 ...
##   ..$ gene_desc: chr [1:50] &quot;voucher ABDN505 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1, 5.8S ribosomal RNA gene, and internal &quot;| __truncated__ &quot;isolate 6 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1, 5.8S ribosomal RNA gene, and internal transc&quot;| __truncated__ &quot;isolate 5 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1, 5.8S ribosomal RNA gene, and internal transc&quot;| __truncated__ &quot;isolate 4 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1, 5.8S ribosomal RNA gene, and internal transc&quot;| __truncated__ ...
##   ..$ acc_no   : chr [1:50] &quot;KM061711&quot; &quot;JQ809322&quot; &quot;JQ809321&quot; &quot;JQ809320&quot; ...
##   ..$ gi_no    : num [1:50] 6.86e+08 3.88e+08 3.88e+08 3.88e+08 3.88e+08 ...
</code></pre>

<pre><code class="r">y &lt;- ncbi_search2(taxa = query, limit = 50)
</code></pre>

<pre><code>## 
## Retrieving data for taxon &#39;pythium&#39;
## 
## 
## Retrieving data for taxon &#39;phytophthora&#39;
## 
## Working on 4797...
## ...retrieving sequence IDs...
## ...retrieving available genes and their lengths...
## ...done.
## Working on 4783...
## ...retrieving sequence IDs...
## ...retrieving available genes and their lengths...
## ...done.
</code></pre>

<pre><code class="r">str(y)
</code></pre>

<pre><code>## List of 2
##  $ :&#39;data.frame&#39;:    50 obs. of  5 variables:
##   ..$ taxon    : chr [1:50] &quot;Pythium dissotocum&quot; &quot;Pythium dissotocum&quot; &quot;Pythium dissotocum&quot; &quot;Pythium dissotocum&quot; ...
##   ..$ length   : num [1:50] 514 861 871 828 862 862 856 874 894 785 ...
##   ..$ gene_desc: chr [1:50] &quot;voucher ABDN358 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1 and 5.8S ribosomal RNA gene, complete s&quot;| __truncated__ &quot;voucher ABDN356 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1, 5.8S ribosomal RNA gene, and internal &quot;| __truncated__ &quot;voucher ABDN355 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1, 5.8S ribosomal RNA gene, and internal &quot;| __truncated__ &quot;voucher ABDN354 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1 and 5.8S ribosomal RNA gene, complete s&quot;| __truncated__ ...
##   ..$ acc_no   : chr [1:50] &quot;KM061708&quot; &quot;KM061706&quot; &quot;KM061705&quot; &quot;KM061704&quot; ...
##   ..$ gi_no    : num [1:50] 6.86e+08 6.86e+08 6.86e+08 6.86e+08 6.86e+08 ...
##  $ :&#39;data.frame&#39;:    50 obs. of  5 variables:
##   ..$ taxon    : chr [1:50] &quot;Phytophthora humicola&quot; &quot;Phytophthora multivora&quot; &quot;Phytophthora multivora&quot; &quot;Phytophthora citrophthora&quot; ...
##   ..$ length   : num [1:50] 899 813 860 840 833 846 874 875 972 782 ...
##   ..$ gene_desc: chr [1:50] &quot;voucher ABDN505 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1, 5.8S ribosomal RNA gene, and internal &quot;| __truncated__ &quot;isolate 6 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1, 5.8S ribosomal RNA gene, and internal transc&quot;| __truncated__ &quot;isolate 5 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1, 5.8S ribosomal RNA gene, and internal transc&quot;| __truncated__ &quot;isolate 4 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1, 5.8S ribosomal RNA gene, and internal transc&quot;| __truncated__ ...
##   ..$ acc_no   : chr [1:50] &quot;KM061711&quot; &quot;JQ809322&quot; &quot;JQ809321&quot; &quot;JQ809320&quot; ...
##   ..$ gi_no    : num [1:50] 6.86e+08 3.88e+08 3.88e+08 3.88e+08 3.88e+08 ...
</code></pre>

<pre><code class="r">all.equal(x, y)
</code></pre>

<pre><code>## [1] TRUE
</code></pre>

<h4>No results</h4>

<p>The two functions differ in how they handle edge cases, like no results or invalid taxon names. My version of the function returns a 0-row dataframe if no sequences are found. This allows for the results of sucessful and unsucessful searches to be <code>rbind</code>ed together without further modifications. On the other hand, an invalid search term returns <code>NA</code>, since this is something the use probably did not intend.</p>

<pre><code class="r">query &lt;- c(&quot;fungi&quot;, &quot;this_is_not_valid&quot;)
x &lt;- ncbi_search(taxa = query, limit = 50)
</code></pre>

<pre><code>## Working on fungi...
## ...retrieving sequence IDs...
## ...retrieving available genes and their lengths...
## ...done.
## Working on this_is_not_valid...
## ...retrieving sequence IDs...
## no sequences for this_is_not_valid - getting other sp.
## no sequences for this_is_not_valid
## ...done.
</code></pre>

<pre><code class="r">str(x)
</code></pre>

<pre><code>## List of 2
##  $ :&#39;data.frame&#39;:    1 obs. of  5 variables:
##   ..$ taxon    : logi NA
##   ..$ length   : logi NA
##   ..$ gene_desc: logi NA
##   ..$ acc_no   : logi NA
##   ..$ gi_no    : logi NA
##  $ :&#39;data.frame&#39;:    1 obs. of  5 variables:
##   ..$ taxon    : Factor w/ 1 level &quot;this_is_not_valid&quot;: 1
##   ..$ length   : logi NA
##   ..$ gene_desc: logi NA
##   ..$ acc_no   : logi NA
##   ..$ gi_no    : logi NA
</code></pre>

<pre><code class="r">y &lt;- ncbi_search2(taxa = query, limit = 50)
</code></pre>

<pre><code>## 
## Retrieving data for taxon &#39;fungi&#39;
## 
## 
## Retrieving data for taxon &#39;this_is_not_valid&#39;
## 
## Not found. Consider checking the spelling or alternate classification
## Working on 4751...
## ...retrieving sequence IDs...
## ...retrieving available genes and their lengths...
## ...done.
</code></pre>

<pre><code class="r">str(y)
</code></pre>

<pre><code>## List of 2
##  $ :&#39;data.frame&#39;:    0 obs. of  5 variables:
##   ..$ taxon    : chr(0) 
##   ..$ length   : num(0) 
##   ..$ gene_desc: chr(0) 
##   ..$ acc_no   : chr(0) 
##   ..$ gi_no    : num(0) 
##  $ : logi NA
</code></pre>

<pre><code class="r">isTRUE(all.equal(x, y))
</code></pre>

<pre><code>## [1] FALSE
</code></pre>

<p>An error during processing still returns <code>NULL</code> for that element, like it does in the current version of <code>ncbi_search</code>. This is a logical output, but there might be cases when a user is running a list of queries and <code>rbind</code>ing the result. In this case, the error could be overlooked, since there would be no record of it in the data. Im not sure what a better choice would be, so I left it as <code>NULL</code>.</p>

<pre><code class="r">all.equal(rbind(NULL, data.frame(1:10)), data.frame(1:10))
</code></pre>

<pre><code>## [1] TRUE
</code></pre>

<h3>New functionality</h3>

<h4><code>entrez_query</code> option</h4>

<p>This option allows advanced users to add their own filter to the esearch query using entrez query syntax. </p>

<pre><code class="r">query &lt;- c(&quot;pythium&quot;)
x &lt;- ncbi_search2(taxa = query, limit = 50, entrez_query = &quot;18S[title] AND 28S[title]&quot;)
</code></pre>

<pre><code>## 
## Retrieving data for taxon &#39;pythium&#39;
## 
## Working on 4797...
## ...retrieving sequence IDs...
## ...retrieving available genes and their lengths...
## ...done.
</code></pre>

<pre><code class="r">head(x$gene_desc)
</code></pre>

<pre><code>## [1] &quot;voucher ABDN356 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1, 5.8S ribosomal RNA gene, and internal transcribed spacer 2, complete sequence; and 28S ribosomal RNA gene, partial sequence&quot;
## [2] &quot;voucher ABDN355 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1, 5.8S ribosomal RNA gene, and internal transcribed spacer 2, complete sequence; and 28S ribosomal RNA gene, partial sequence&quot;
## [3] &quot;voucher ABDN353 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1, 5.8S ribosomal RNA gene, and internal transcribed spacer 2, complete sequence; and 28S ribosomal RNA gene, partial sequence&quot;
## [4] &quot;voucher ABDN352 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1, 5.8S ribosomal RNA gene, and internal transcribed spacer 2, complete sequence; and 28S ribosomal RNA gene, partial sequence&quot;
## [5] &quot;voucher ABDN351 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1, 5.8S ribosomal RNA gene, and internal transcribed spacer 2, complete sequence; and 28S ribosomal RNA gene, partial sequence&quot;
## [6] &quot;voucher ABDN350 18S ribosomal RNA gene, partial sequence; internal transcribed spacer 1, 5.8S ribosomal RNA gene, and internal transcribed spacer 2, complete sequence; and 28S ribosomal RNA gene, partial sequence&quot;
</code></pre>

<h4><code>hypothetical</code> option</h4>

<p>This allows users to include hypothetical/predicted genes if they want. The default to remove them, as in the built-in functionality of the current <code>ncbi_search</code>. </p>

<pre><code class="r">query &lt;- c(&quot;fungi&quot;)
x &lt;- ncbi_search2(taxa = query, limit = 50)
</code></pre>

<pre><code>## 
## Retrieving data for taxon &#39;fungi&#39;
## 
## Working on 4751...
## ...retrieving sequence IDs...
## ...retrieving available genes and their lengths...
## ...done.
</code></pre>

<pre><code class="r">str(x)
</code></pre>

<pre><code>## &#39;data.frame&#39;:    0 obs. of  5 variables:
##  $ taxon    : chr 
##  $ length   : num 
##  $ gene_desc: chr 
##  $ acc_no   : chr 
##  $ gi_no    : num
</code></pre>

<pre><code class="r">y &lt;- ncbi_search2(taxa = query, limit = 50, hypothetical = TRUE)
</code></pre>

<pre><code>## 
## Retrieving data for taxon &#39;fungi&#39;
## 
## Working on 4751...
## ...retrieving sequence IDs...
## ...retrieving available genes and their lengths...
## ...done.
</code></pre>

<pre><code class="r">str(y)
</code></pre>

<pre><code>## &#39;data.frame&#39;:    50 obs. of  5 variables:
##  $ taxon    : chr  &quot;Wallemia ichthyophaga&quot; &quot;Wallemia ichthyophaga&quot; &quot;Wallemia ichthyophaga&quot; &quot;Wallemia ichthyophaga&quot; ...
##  $ length   : num  1392 1299 441 615 2682 ...
##  $ gene_desc: chr  &quot;EXF-994 SWI/SNF-related matrix-associated actin-dependent regulator of chromatin subfamily D member 2 partial mRNA&quot; &quot;EXF-994 hypothetical protein partial mRNA&quot; &quot;EXF-994 22.0 kDa heat shock protein partial mRNA&quot; &quot;EXF-994 hypothetical protein partial mRNA&quot; ...
##  $ acc_no   : chr  &quot;009267788&quot; &quot;009267787&quot; &quot;009267786&quot; &quot;009267785&quot; ...
##  $ gi_no    : num  6.86e+08 6.86e+08 6.86e+08 6.86e+08 6.86e+08 ...
</code></pre>

</body>

</html>

